### PR TITLE
add model tag

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

Add model tag in serialized message

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by wzhongyuan** [comment]
@jason-dai  @SeaOfOcean   Please help review the change

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -473,6 +473,22 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+  /**
+   * Save this module with class tag to path with protobuf format
+   * @param path path to save module, local file system, HDFS and Amazon S3 is supported.
+   *             HDFS path should be like "hdfs://[host]:[port]/xxx"
+   *             Amazon S3 path should be like "s3a://bucket/xxx"
+   * @param moduleTag  module tag
+   * @param overWrite if overwrite
+   * @return self
+   */
+  def saveModuleWithTag(path : String, moduleTag : ModuleTag, overWrite
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Why do we need a new API? Just make moduleTag an optional argument in saveModule

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Module.scala [hunk 2]
```diff
@@ -58,6 +58,20 @@ object Module {
...
+  /**
+   * Load model from path.
+   *
+   * @param path path to save module, local file system, HDFS and Amazon S3 is supported.
+   *             HDFS path should be like "hdfs://[host]:[port]/xxx"
+   *             Amazon S3 path should be like "s3a://bucket/xxx"
+   * @tparam T numeric type
+   * @return model loaded from path
+   */
+  def loadModuleWithTag[T: ClassTag](path : String)(implicit ev: TensorNumeric[T])
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Module.scala]: Don't add a new API; just add an optional argument for storing tag to loadModule.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializer.scala [hunk 3]
```diff
@@ -211,3 +211,9 @@ private case class GenericTypeWrapper[T: ClassTag](tensor : Tensor[T],
   ttpe : T
   )(implicit ev: TensorNumeric[T])
 
+// ModelTag is an unique identifier for a serialized model
+case class ModuleTag(publisher : String,
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializer.scala]: Do we need a ModuleTag class? Or maybe just use a string which can contain arbitrary tag, e.g., bigdl.v_0_3_0.ssd?

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializer.scala]: I discussed with @SeaOfOcean ,  Maybe with a ModuleTag we could handle the compatibility in a more elegant way in application (model_zoo or SSD model)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Module.scala (hunk 2 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Module.scala]: @jason-dai   The tag is not a class field of 'AbstractModule', if we use the same API,  we may want to add that in AbstractModule, shall we ?

**Comment by wzhongyuan** [comment]
As discussed offline, no need to add a separate tag, closing this one

**Pull request closed without merging**